File: ros2_ws/src/action_demo_py/action_demo_py/dishes_client.py
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node

from action_demo_interfaces.action import DoDishes


class DoDishesClient(Node):
    def __init__(self) -> None:
        super().__init__("dishes_client")
        self._client = ActionClient(self, DoDishes, "dishes")
        self._done = False
        self._exit_code = 0

    def run(self) -> int:
        if not self._client.wait_for_server(timeout_sec=5.0):
            self.get_logger().error("Action server not available")
            return 1

        goal_msg = DoDishes.Goal()
        goal_msg.dishwasher_id = 2

        future = self._client.send_goal_async(
            goal_msg, feedback_callback=self.feedback_callback
        )
        future.add_done_callback(self.goal_response_callback)

        while rclpy.ok() and not self._done:
            rclpy.spin_once(self, timeout_sec=0.1)

        return self._exit_code

    def feedback_callback(self, feedback_msg) -> None:
        self.get_logger().info(
            f"Feedback {feedback_msg.feedback.percent_complete:.1f}%"
        )

    def goal_response_callback(self, future) -> None:
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().error("Goal was rejected by server")
            self._done = True
            self._exit_code = 1
            return

        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(self.result_callback)

    def result_callback(self, future) -> None:
        result = future.result().result
        self.get_logger().info(
            f"Yay! The dishes are now clean. Total cleaned: {result.total_dishes_cleaned}"
        )
        self._done = True


def main() -> None:
    rclpy.init()
    node = DoDishesClient()
    try:
        raise SystemExit(node.run())
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: ros2_ws/src/action_demo_cpp/src/dishes_client.cpp
#include <chrono>
#include <cstdlib>
#include <future>
#include <memory>

#include "action_demo_interfaces/action/do_dishes.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

using namespace std::chrono_literals;

class DoDishesClient : public rclcpp::Node
{
public:
  using DoDishes = action_demo_interfaces::action::DoDishes;
  using GoalHandleDoDishes = rclcpp_action::ClientGoalHandle<DoDishes>;

  DoDishesClient()
  : Node("dishes_client")
  {
    client_ = rclcpp_action::create_client<DoDishes>(this, "dishes");
  }

  int run()
  {
    if (!client_->wait_for_action_server(5s)) {
      RCLCPP_ERROR(get_logger(), "Action server not available");
      return 1;
    }

    DoDishes::Goal goal_msg;
    goal_msg.dishwasher_id = 2;

    auto send_goal_options = rclcpp_action::Client<DoDishes>::SendGoalOptions();
    send_goal_options.feedback_callback =
      [this](GoalHandleDoDishes::SharedPtr, const std::shared_ptr<const DoDishes::Feedback> feedback) {
        RCLCPP_INFO(get_logger(), "Feedback %.1f%%", feedback->percent_complete);
      };

    auto goal_handle_future = client_->async_send_goal(goal_msg, send_goal_options);
    if (rclcpp::spin_until_future_complete(shared_from_this(), goal_handle_future) !=
      rclcpp::FutureReturnCode::SUCCESS)
    {
      RCLCPP_ERROR(get_logger(), "Failed to send goal");
      return 1;
    }

    auto goal_handle = goal_handle_future.get();
    if (!goal_handle) {
      RCLCPP_ERROR(get_logger(), "Goal was rejected by server");
      return 1;
    }

    auto result_future = client_->async_get_result(goal_handle);
    if (rclcpp::spin_until_future_complete(shared_from_this(), result_future) !=
      rclcpp::FutureReturnCode::SUCCESS)
    {
      RCLCPP_ERROR(get_logger(), "Failed to get result");
      return 1;
    }

    const auto wrapped_result = result_future.get();
    if (wrapped_result.code != rclcpp_action::ResultCode::SUCCEEDED) {
      RCLCPP_ERROR(get_logger(), "Action did not succeed");
      return 1;
    }

    RCLCPP_INFO(
      get_logger(),
      "Yay! The dishes are now clean. Total cleaned: %u",
      wrapped_result.result->total_dishes_cleaned);
    return 0;
  }

private:
  rclcpp_action::Client<DoDishes>::SharedPtr client_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<DoDishesClient>();
  const int rc = node->run();
  rclcpp::shutdown();
  return rc;
}
